File: src/listener.rs
use crate::state::{Message, SharedState};
use sqlx::postgres::PgListener;
use tokio_postgres::Error;

pub async fn listener(state: SharedState) -> Result<(), Error> {
    let mut listener = PgListener::connect("postgres://postgres:postgres@localhost:5432/postgres")
        .await
        .unwrap();

    tokio::task::spawn(async move {
        let _ = listener.listen("test_notifications").await;
        loop {
            let message = listener.recv().await;

            match message {
                Ok(m) => {
                    println!("{}", m.channel());
                    let state = state.read().await;
                    let clients = &state.clients;

                    for (_, sender) in clients {
                        let _ = sender.send(Message::Text(m.payload().to_string())).await;
                    }
                }
                Err(e) => eprintln!("{}", e),
            }
        }
    });

    Ok(())
}

File: src/listener.ts
import { Client } from "pg";
import { clients } from "./state.js";

export async function startListener(): Promise<void> {
  const pgClient = new Client({
    host: "localhost",
    port: 5432,
    database: "postgres",
    user: "postgres",
    password: "postgres",
  });

  await pgClient.connect();
  console.log("PostgreSQL connected");

  await pgClient.query("LISTEN test_notifications");

  pgClient.on("notification", (notification) => {
    console.log(`Received notification: ${notification.channel} - ${notification.payload}`);

    for (const client of clients.values()) {
      if (client.ws.readyState === 1) {
        client.ws.send(notification.payload ?? "" );
      }
    }
  });

  pgClient.on("error", (err) => {
    console.error("PostgreSQL error:", err);
  });
}
